File: src/main/java/org/dbtools/query/shared/filter/InFilter.java
package org.dbtools.query.shared.filter;

import org.dbtools.query.shared.QueryBuilder;

import javax.annotation.Nonnull;
import java.util.List;

public class InFilter extends CompareFilter {

    protected boolean in;
    private QueryBuilder subQuery;

    public static InFilter create(String field, Object value) {
        InFilter filter = new InFilter();
        filter.filter = InFilter.newInstance(field, true, value);
        return filter;
    }

    public static InFilter create(String field, boolean in, Object values) {
        InFilter filter = new InFilter();
        filter.filter = InFilter.newInstance(field, in, values);
        return filter;
    }

    public static InFilter create(String field, QueryBuilder subQuery) {
        InFilter filter = new InFilter();
        filter.filter = InFilter.newInstance(field, true, subQuery);
        return filter;
    }

    public static InFilter create(String field, boolean in, QueryBuilder subQuery) {
        InFilter filter = new InFilter();
        filter.filter = InFilter.newInstance(field, in, subQuery);
        return filter;
    }

    private static InFilter newInstance(String field, boolean in, Object values) {
        return new InFilter(field, in, values);
    }

    private static InFilter newInstance(String field, boolean in, QueryBuilder subQuery) {
        return new InFilter(field, in, subQuery);
    }

    protected InFilter() {
        super();
    }

    protected InFilter(String field, boolean in, Object value) {
        super(field, null, value);
        if (value instanceof List && ((List) value).size() < 1) {
            throw new IllegalArgumentException("List must contain at least on item");
        }
        this.in = in;
        this.subQuery = null;
    }

    private InFilter(String field, boolean in, QueryBuilder subQuery) {
        super(field, null, null);
        this.in = in;
        this.subQuery = subQuery;
    }

    @Override
    public String build(@Nonnull QueryBuilder queryBuilder) {
        StringBuilder builder = new StringBuilder(field);
        if (in) {
            builder.append(" IN ");
        } else {
            builder.append(" NOT IN ");
        }
        builder.append("(");
        if (subQuery != null) {
            builder.append(subQuery.buildQuery());
        } else if (value instanceof List) {
            List list = (List) value;
            builder.append(queryBuilder.formatValue(list.get(0)));
            int count = list.size();
            for (int i = 1; i < count; i++) {
                builder.append(", ").append(queryBuilder.formatValue(list.get(i)));
            }
        } else {
            builder.append(queryBuilder.formatValue(value));
        }
        return builder.append(")").toString();
    }

    public InFilter and(String field, Object value) {
        and(InFilter.newInstance(field, true, value));
        return this;
    }

    public InFilter or(String field, Object value) {
        or(InFilter.newInstance(field, true, value));
        return this;
    }

    public InFilter and(String field, boolean in, Object value) {
        and(InFilter.newInstance(field, in, value));
        return this;
    }

    public InFilter or(String field, boolean in, Object value) {
        or(InFilter.newInstance(field, in, value));
        return this;
    }

    public InFilter and(String field, QueryBuilder queryBuilder) {
        and(InFilter.newInstance(field, true, queryBuilder));
        return this;
    }

    public InFilter or(String field, QueryBuilder queryBuilder) {
        or(InFilter.newInstance(field, true, queryBuilder));
        return this;
    }

    public InFilter and(String field, boolean in, QueryBuilder queryBuilder) {
        and(InFilter.newInstance(field, in, queryBuilder));
        return this;
    }

    public InFilter or(String field, boolean in, QueryBuilder queryBuilder) {
        or(InFilter.newInstance(field, in, queryBuilder));
        return this;
    }



    @Override
    public InFilter clone() {
        InFilter clone = (InFilter) super.clone();
        clone.in = this.in;
        clone.subQuery = this.subQuery;
        return clone;
    }
}

File: dbtools-query/src/commonMain/kotlin/org/dbtools/query/shared/filter/InFilter.kt
package org.dbtools.query.shared.filter

import org.dbtools.query.shared.QueryBuilder

open class InFilter : CompareFilter {
    protected var isIn: Boolean = false
    private var subQuery: QueryBuilder? = null

    constructor() : super()

    constructor(field: String?, isIn: Boolean, value: Any? = null) : super(field, null, value) {
        require(!(value is List<*> && value.size < 1)) { "List must contain at least on item" }
        this.isIn = isIn
        this.subQuery = null
    }

    constructor(field: String?, isIn: Boolean, subQuery: QueryBuilder) : super(field, null, null) {
        this.isIn = isIn
        this.subQuery = subQuery
    }

    override fun build(queryBuilder: QueryBuilder): String = buildString {
        append(field)
        if (isIn) {
            append(" IN ")
        } else {
            append(" NOT IN ")
        }
        append("(")
        val currentSubQuery = subQuery
        if (currentSubQuery != null) {
            append(currentSubQuery.buildQuery())
        } else if (value is List<*>) {
            val list = value as List<*>
            append(queryBuilder.formatValue(list[0]))
            val count = list.size
            for (i in 1 until count) {
                append(", ").append(queryBuilder.formatValue(list[i]))
            }
        } else {
            append(queryBuilder.formatValue(value))
        }
        return append(")").toString()
    }

    override fun and(field: String, value: Any): InFilter {
        and(newInstance(field, true, value))
        return this
    }

    override fun or(field: String, value: Any): InFilter {
        or(newInstance(field, true, value))
        return this
    }

    fun and(field: String, isIn: Boolean, value: Any): InFilter {
        and(newInstance(field, isIn, value))
        return this
    }

    fun or(field: String, isIn: Boolean, value: Any): InFilter {
        or(newInstance(field, isIn, value))
        return this
    }

    fun and(field: String, queryBuilder: QueryBuilder): InFilter {
        and(newInstance(field, true, queryBuilder))
        return this
    }

    fun or(field: String, queryBuilder: QueryBuilder): InFilter {
        or(newInstance(field, true, queryBuilder))
        return this
    }

    fun and(field: String, isIn: Boolean, queryBuilder: QueryBuilder): InFilter {
        and(newInstance(field, isIn, queryBuilder))
        return this
    }

    fun or(field: String, isIn: Boolean, queryBuilder: QueryBuilder): InFilter {
        or(newInstance(field, isIn, queryBuilder))
        return this
    }

    companion object {
        fun create(field: String, value: Any): InFilter {
            val filter = InFilter()
            filter.filter = newInstance(field, true, value)
            return filter
        }

        fun create(field: String, isIn: Boolean, values: Any): InFilter {
            val filter = InFilter()
            filter.filter = newInstance(field, isIn, values)
            return filter
        }

        fun create(field: String, subQuery: QueryBuilder): InFilter {
            val filter = InFilter()
            filter.filter = newInstance(field, true, subQuery)
            return filter
        }

        fun create(field: String, isIn: Boolean, subQuery: QueryBuilder): InFilter {
            val filter = InFilter()
            filter.filter = newInstance(field, isIn, subQuery)
            return filter
        }

        private fun newInstance(field: String, isIn: Boolean, values: Any): InFilter {
            return InFilter(field, isIn, values)
        }

        private fun newInstance(field: String, isIn: Boolean, subQuery: QueryBuilder): InFilter {
            return InFilter(field, isIn, subQuery)
        }
    }
}
